solver/src/settlement.rs

```diff
@@ -1,27 +1,36 @@
-use model::{Order, TokenPair, UserOrder};
-use std::{
-    collections::HashMap,
-    time::{Duration, Instant},
-};
+use model::UserOrder;
+use primitive_types::{H160, U256};
+use std::collections::HashMap;
 
-pub struct Settlement {
-    pub token_pair: TokenPair,
-    pub orders: Vec<UserOrder>,
-    // TODO: sell amounts, AMM trades
+#[derive(Debug)]
+pub struct Trade {
+    pub order: UserOrder,
+    pub executed_amount: U256,
+}
+
+#[derive(Debug)]
+pub enum Interaction {
+    // https://uniswap.org/docs/v2/smart-contracts/router02/#swapexacttokensfortokens
+    UniswapExactTokensForTokens {
+        amount_in: U256,
+        amount_out_min: U256,
+        token_in: H160,
+        token_out: H160,
+    },
+    // https://uniswap.org/docs/v2/smart-contracts/router02/#swaptokensforexacttokens
+    UniswapTokensForExactTokens {
+        amount_out: U256,
+        amount_in_max: U256,
+        token_in: H160,
+        token_out: H160,
+    },
 }
 
-// Assumes all orders have spending approved and enough balance.
-pub fn find_settlement(
-    _orders: &[Order],
-    _nonces: &HashMap<TokenPair, u32>,
-    _now: Instant,
-    _max_order_age: Duration,
-    // TODO: AMM price
-) -> Option<Settlement> {
-    // TODO: Find oldest order where nonce is correct. If it is older than max_order_age settle that pair.
-    // TODO: implement solving algorithm
-    // Not sure how we should handle the AMM prices for token pairs. If a token pair only has
-    // unreasonably priced orders then we cannot settle it. But there could be a counter
-    // unreasonable order that does allow settling.
-    todo!()
+#[derive(Debug, Default)]
+pub struct Settlement {
+    pub clearing_prices: HashMap<H160, U256>,
+    pub fee_factor: U256,
+    pub trades: Vec<Trade>,
+    pub interactions: Vec<Interaction>,
+    pub order_refunds: Vec<()>,
```

Comment:
what is this for?

Reply:
I took it from the smart contract interface:
>     /// @param encodedOrderRefunds Encoded order refunds for clearing storage

Reply:
I see. @nlordell I was hoping we could model storage freeings as an interaction as well (rather than bloating up the public interface).

Reply:
I think that is a really good idea.